api/v1alpha1/release_types_test.go

```diff
@@ -0,0 +1,671 @@
+					Conditions: []metav1.Condition{
+						{
+							Type:   releaseConditionType,
+							Status: metav1.ConditionUnknown,
+						},
+					},
+				},
+			}
+			Expect(r.HasSucceeded()).To(BeFalse(),
+				"Expected to get: %+v ; Actually Got: %+v\n", false, true)
+		})
+
+		It("should return true because condition.Status is ConditionTrue", func() {
+			r := &Release{
+				Status: ReleaseStatus{
+					Conditions: []metav1.Condition{
+						{
+							Type:   releaseConditionType,
+							Status: metav1.ConditionTrue,
+						},
+					},
+				},
+			}
+			Expect(r.HasSucceeded()).To(BeTrue(),
+				"Expected to get: %+v ; Actually Got: %+v\n", true, false)
+		})
+	})
+
+	Context("IsDone()", func() {
+		It("should return false because condition struct is nil", func() {
+			r := &Release{}
+			Expect(r.IsDone()).To(BeFalse(),
+				"Expected to get: %+v ; Actually Got: %+v\n", false, true)
+		})
+
+		It("should return true because condition.Status is ConditionTrue", func() {
+			r := &Release{
+				Status: ReleaseStatus{
+					Conditions: []metav1.Condition{
+						{
+							Type:   releaseConditionType,
+							Status: metav1.ConditionTrue,
+						},
+					},
+				},
+			}
+			Expect(r.IsDone()).To(BeTrue(),
+				"Expected to get: %+v ; Actually Got: %+v\n", true, false)
+		})
+
+		It("should return false because condition.Status is ConditionUnknown", func() {
+			r := &Release{
+				Status: ReleaseStatus{
+					Conditions: []metav1.Condition{
+						{
+							Type:   releaseConditionType,
+							Status: metav1.ConditionUnknown,
+						},
+					},
+				},
+			}
+			Expect(r.IsDone()).To(BeFalse(),
+				"Expected to get: %+v ; Actually Got: %+v\n", false, true)
+		})
+	})
+
+	Context("MarkFailed()", func() {
+		It("should NOOP because IsDone() is true, and Status.CompletionTime is not nil", func() {
+			args := commonArgs{
+				reason:  ReleaseReasonFoo,
+				message: "cow say m00",
+			}
+			r := &Release{
+				Status: ReleaseStatus{
+					StartTime: &metav1.Time{
+						Time: time.Now().Add(-time.Hour * 1),
+					},
+					CompletionTime: &metav1.Time{ // not nil
+						Time: time.Now(),
+					},
+					Conditions: []metav1.Condition{
+						{
+							Type:    releaseConditionType,
+							Status:  metav1.ConditionTrue,
+							Reason:  ReleaseReasonBar.String(),
+							Message: "Testcase one message string",
+						},
+					},
+				},
+			}
+			r.MarkFailed(args.reason, args.message)
+			Expect(len(r.Status.Conditions)).To(
+				Equal(1),
+				"Expected number of []Conditions to equal \"%+v\"\n",
+				1)
+			Expect(r.Status.Conditions[0]).To(MatchFields(IgnoreMissing|IgnoreExtras, Fields{
+				"Status":  Equal(metav1.ConditionTrue),
+				"Type":    Equal(releaseConditionType),
+				"Reason":  Equal(ReleaseReasonBar.String()),
+				"Message": Equal("Testcase one message string"),
+			}), "Expect condition struct fields to match expected fields")
+		})
+
+		It("should not NOOP because CompletionTime is nil, and IsDone returns false", func() {
```

Comment:
This description sound a bit confusing to me.

Reply:
I agree it's a bit wordy. I've tried to clarify a little, and added a truth table help.